RBAC: protect folder creation and moving

IevaVasiljeva opened the pull request:
**What is this feature?**

Protect nested folder creation and moving:
* users can only create folder under another folder if they have the permissions to create a folder and write access to this other folder;
* users can only move folder to another folder if they have write access to both folders.

Also refactored folder API tests to not use RBAC mock anymore and added some extra tests that we had missed (for folder creation and updates).

**Which issue(s) does this PR fix?**:

Related to https://github.com/grafana/grafana-authnz-team/issues/172

**Special notes for your reviewer**:

IevaVasiljeva committed:
protect moving folders to a subfolder and creating folders in a subfolder

IevaVasiljeva commented on pkg/api/folder.go:
```diff
@@ -186,7 +198,7 @@ func (hs *HTTPServer) setDefaultFolderPermissions(ctx context.Context, orgID int
 		}
 
-		if !isNested {
+		if !isNested || !hs.Features.IsEnabled(featuremgmt.FlagNestedFolders) {
```
Small change to make sure that we don't set permissions wrongly if parent folder for some reason is set while nested folders are not enabled.

IevaVasiljeva commented on pkg/api/folder.go:
```diff
@@ -210,8 +222,21 @@ func (hs *HTTPServer) MoveFolder(c *contextmodel.ReqContext) response.Response {
 		var theFolder *folder.Folder
 		var err error
 		if cmd.NewParentUID != "" {
+			if !hs.AccessControl.IsDisabled() {
```
question: should we check that the user has `folders:create` permissions if they're trying to move the folder to the root?

Jguer replied:
that would make sense as it would be a workaround to bypass not having `folders:create`, have you seen any downsides to it?

kalleep replied:
Yeah I think that make sense

IevaVasiljeva committed:
folder update endpoint isn't used for folder parent update

IevaVasiljeva committed:
lint

kalleep reviewed (commented):
Look good! I think we should check "folders:create" when moving too root. Also looking at the code we are now performing access control checks in middleware, api handler and in service. I guess we should clean that up

IevaVasiljeva committed:
merge main

IevaVasiljeva committed:
move permission check logic to services, fix tests

IevaVasiljeva commented on pkg/api/folder.go:
```diff
@@ -151,6 +151,18 @@ func (hs *HTTPServer) CreateFolder(c *contextmodel.ReqContext) response.Response
 
+	if !hs.AccessControl.IsDisabled() && hs.Features.IsEnabled(featuremgmt.FlagNestedFolders) && cmd.ParentUID != "" {
+		// Check that the user is allowed to create a subfolder in this folder
+		evaluator := accesscontrol.EvalPermission(dashboards.ActionFoldersWrite, dashboards.ScopeFoldersProvider.GetResourceScopeUID(cmd.ParentUID))
```
I think there'll be a separate guard preventing users creating folders under the general folder. There's even a chance that general folder will go away completely. But I'll make a note to ask that in the next nested folder sync.

IevaVasiljeva commented:
> Look good! I think we should check "folders:create" when moving too root. Also looking at the code we are now performing access control checks in middleware, api handler and in service. I guess we should clean that up

I've refactored the code to do all the checks from the folder service. It's still pretty messy because there's guardians and also direct access control permission evaluations, but I think that's something that we can clean up post Grafana 10 when we've removed the option to disable RBAC.

I've also added `folders:create` check for moving to the root. It feels a bit weird (different actions required for different moves), but better safe than sorry I guess.

IevaVasiljeva committed:
linting

papagian reviewed (approved):
LGTM
